core/controllers/question_player.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2019 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Controllers for the questions player."""
+
+from core.controllers import acl_decorators
+from core.controllers import base
+from core.domain import skill_domain
+from core.domain import skill_services
+
+
+class SkillMasteryDataHandler(base.BaseHandler):
+    """A handler that handles fetching and updating the degrees of user
+    skill mastery.
+    """
+
+    @acl_decorators.can_access_user_skill_mastery
+    def put(self):
+        """Handles PUT requests."""
+        degree_of_mastery_per_skill = (
+            self.payload.get('degree_of_mastery_per_skill'))
+        if (not degree_of_mastery_per_skill or
+                not isinstance(degree_of_mastery_per_skill, dict)):
+            raise self.PageNotFoundException
+
+        skill_ids = degree_of_mastery_per_skill.keys()
+
+        try:
+            for skill_id in skill_ids:
+                skill_domain.Skill.require_valid_skill_id(skill_id)
+        except Exception:
+            raise self.InvalidInputException
+
+        try:
+            skill_services.get_multi_skills(skill_ids)
+        except Exception as e:
+            raise self.PageNotFoundException(e)
+
+        for skill_id in skill_ids:
```

Comment:
You can perform these checks in the for loop above, to avoid iterating through the skill_ids list twice.

Reply:
I think we should check valid skill id before checking the skill exists or not.

Reply:
Oh yes I agree, I meant to move these checks above the skill exists check (in the same loop as the skill validation check)

Reply:
Sorry I'm still a little confused. Did you mean moving the checks for type and value of degree_of_mastery before the skill exists check? But the checks would be unnecessary if the skill doesn't exist, so I don't think it's efficient to do it in that way.

Reply:
So the chances of sending skill ids where skills don't exist are quite low. You will most of the time pass the skill exist check. So what will happen is that most of the time you will iterate through skill_ids twice compared to once. That turns out to be more inefficient compared to doing all data related checks in one loop.

Reply:
Ok that makes sense. I have changed it.